stdlib/public/runtime/MetadataLookup.cpp

```diff
@@ -243,10 +247,112 @@ swift::_matchMetadataByMangledTypeName(const llvm::StringRef typeName,
   return metadata;
 }
 
-// returns the type metadata for the type named by typeName
+// returns the generic type metadata for type named by boundGenericNode
+static const Metadata *
+matchMetadataGenericNode(const llvm::StringRef typeNodeName,
+                         const NodePointer boundGenericNode,
+                         const Metadata *unused __attribute__((unused)),
+                         const NominalTypeDescriptor *ntd) {
+  if (boundGenericNode->getNumChildren() != 2)
+    return nullptr;
+
+  auto templateTypeNode = boundGenericNode->getChild(0);
+  auto typeListNode = boundGenericNode->getChild(1);
+
+  if (templateTypeNode->getKind() != Node::Kind::Type ||
+      typeListNode->getKind() != Node::Kind::TypeList)
+    return nullptr;
+
+  // Validate that we have one or more generic parameters (the zero parameter
+  // case is handled by matchMetadataTypeNode) and that the number of primary
+  // archetypes matches the number of child nodes in the demangled type.
+  //
+  // TODO: support secondary archetypes, where NumParams != NumPrimaryParams;
```

Comment:
@rjmccall is working on redoing how generic parameter lists are represented in type metadata and the calling convention so that passing secondary archetypes will not be necessary; instead, they will be stored in witness tables.

You'll still need to do checks to ensure the associated types conform to the requirements of the type's generic signature, though.